feat: add C implementation for `stats/base/dists/signrank/pdf`

Lokeshranjan8 opened the pull request:
Resolves #3884 

## Description

> What is the purpose of this pull request?

This pull request:

-     adds C implementation for stats/base/dists/signrank/pdf along with relevant docs, tests, examples and benchmarks

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3884 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Lokeshranjan8 committed:
feat: add c implementation for stats/base/dists/signrank/pdf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
chore: minor clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
refactor: type n as an int32

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/stats/base/dists/signrank/pdf.h"
+#include "stdlib/math/base/napi/binary.h"
+
+// cppcheck-suppress shadowFunction
```
```suggestion
```

Lokeshranjan8 committed:
Apply suggestion from @Planeshifter

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/benchmark/benchmark.js:
```diff
@@ -20,53 +20,44 @@
-	var x;
-	var y;
-	var i;
+var pdf = tryRequire( resolve( __dirname, './../lib/native.js' ) );
```
The contents of this file should be `benchmark.native.js`, while what is currently in `benchmark.native.js` should become `benchmark.js`.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/src/main.c:
```diff
@@ -0,0 +1,99 @@
+* Calculates the weight for the `(x,n)` pair and memoizes the result.
+*
+* @private
+* @param {number} x - input value
```
Need to replace the JSDoc comments with C Doxygen style comments here and below.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/src/main.c:
```diff
@@ -0,0 +1,99 @@
+* @param {PositiveInteger} n - number of observations
+* @returns {Probability} evaluated PDF
+*
+* @example
+* var y = pdf( 7.0, 9 );
+* // returns ~0.01
+*
+* @example
+* var y = pdf( 7.0, 6 );
+* // returns ~0.063
+*
+* @example
+* var y = pdf( -1.0, 40 );
+* // returns 0.0
+*
+* @example
+* var y = pdf( NaN, 10 );
+* // returns NaN
+*
+* @example
+* var y = pdf( 0.0, NaN );
+* // returns NaN
+*
+* @example
+* var y = pdf( 2.0, -1 );
+* // returns NaN
+*
+* @example
+* var y = pdf( 2.0, 1.8 );
+* // returns NaN
```
We only keep one example in `main.c` by convention:

```suggestion
* @example
* double y = stdlib_base_dists_signrank_pdf( 7.0, 9 );
* // returns ~0.01
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/manifest.json:
```diff
@@ -0,0 +1,82 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-positive-integer",
+        "@stdlib/math/base/assert/is-finite",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/exp",
+        "@stdlib/constants/float64/ln-two",
+        "@stdlib/math/base/special/ln",
+        "@stdlib/math/base/assert/is-integer"
```
```suggestion
        "@stdlib/stdlib/math/base/special/ceil",
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/math/base/special/exp",
        "@stdlib/constants/float64/ln-two",
        "@stdlib/math/base/special/ln",
        "@stdlib/math/base/assert/is-integer"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/manifest.json:
```diff
@@ -0,0 +1,82 @@
+        "@stdlib/math/base/special/exp",
+        "@stdlib/constants/float64/ln-two",
+        "@stdlib/math/base/special/ln",
+        "@stdlib/math/base/assert/is-integer"
```
```suggestion
        "@stdlib/stdlib/math/base/special/ceil",
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/math/base/special/exp",
        "@stdlib/constants/float64/ln-two",
        "@stdlib/math/base/special/ln",
        "@stdlib/math/base/assert/is-integer"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/manifest.json:
```diff
@@ -0,0 +1,82 @@
+        "@stdlib/constants/float64/ln-two",
+        "@stdlib/math/base/napi/binary",
+        "@stdlib/math/base/special/ln",
+        "@stdlib/math/base/assert/is-integer"
```
```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/math/base/special/exp",
        "@stdlib/constants/float64/ln-two",
        "@stdlib/math/base/napi/binary",
        "@stdlib/math/base/special/ln",
        "@stdlib/math/base/assert/is-integer"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/lib/native.js:
```diff
@@ -0,0 +1,70 @@
+* @example
+* var y = pdf( 2.0, -1 );
+* // returns NaN
+*
+* @example
+* var y = pdf( 2.0, 1.8 );
+* // returns NaN
```
Should remove this example since function now returns `0` for it:

```suggestion
```

Lokeshranjan8 committed:
Apply suggestion from @Planeshifter

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/manifest.json

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/manifest.json

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/manifest.json

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/lib/native.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
fix: implement C Doxygen style comments in main.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Lokeshranjan8 committed:
Update benchmark.native.js

Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Planeshifter committed:
chore: minor clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/signrank/pdf/benchmark/benchmark.js:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2020 The Stdlib Authors.
+* Copyright (c) 2025 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2020 The Stdlib Authors.
```

Planeshifter committed:
chore: keep original copyright year

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
LGTM! Thanks, @Lokeshranjan8!
